internal/file/file_manager_service.go

```diff
@@ -142,3 +158,108 @@ func (fms *FileManagerService) UpdateFile(
 
 	return err
 }
+
+func (fms *FileManagerService) ConfigApply(ctx context.Context, configApplyRequest *mpi.ConfigApplyRequest) error {
+	fileOverview := configApplyRequest.GetOverview()
+
+	if fileOverview == nil {
+		return fmt.Errorf("fileOverview is nil")
+	}
+
+	allowedErr := fms.checkAllowedDirectory(fileOverview.GetFiles())
+	if allowedErr != nil {
+		return allowedErr
+	}
+
+	diffFiles, fileContent, err := files.CompareFileHash(fileOverview)
+	if err != nil {
+		return err
+	}
+
+	fms.fileContentsCache = fileContent
+	fms.filesCache = diffFiles
+
+	fileErr := fms.fileActions(ctx)
+	if fileErr != nil {
+		return fileErr
+	}
+
+	return nil
+}
+
+func (fms *FileManagerService) fileActions(ctx context.Context) error {
+	for _, file := range fms.filesCache {
+		switch file.GetAction() {
+		case mpi.File_FILE_ACTION_DELETE:
+			if err := os.Remove(file.GetFileMeta().GetName()); err != nil && !os.IsNotExist(err) {
+				return fmt.Errorf("error deleting file: %s error: %w", file.GetFileMeta().GetName(), err)
+			}
+
+			continue
+		case mpi.File_FILE_ACTION_ADD, mpi.File_FILE_ACTION_UPDATE:
+			updateErr := fms.fileUpdate(ctx, file)
+			if updateErr != nil {
+				return updateErr
+			}
+		case mpi.File_FILE_ACTION_UNSPECIFIED, mpi.File_FILE_ACTION_UNCHANGED:
+			fallthrough
+		default:
+			slog.DebugContext(ctx, "File Action not implemented")
+		}
+	}
+
+	return nil
+}
+
+func (fms *FileManagerService) fileUpdate(ctx context.Context, file *mpi.File) error {
+	getFileResp, getFileErr := fms.fileServiceClient.GetFile(ctx, &mpi.GetFileRequest{
+		MessageMeta: &mpi.MessageMeta{
+			MessageId:     uuid.NewString(),
+			CorrelationId: logger.GetCorrelationID(ctx),
+			Timestamp:     timestamppb.Now(),
+		},
+		FileMeta: file.GetFileMeta(),
+	})
+	if getFileErr != nil {
+		return fmt.Errorf("error getting file data for %s: %w", file.GetFileMeta(), getFileErr)
+	}
+
+	writeErr := fms.fileOperator.Write(ctx, getFileResp.GetContents().GetContents(), file.GetFileMeta())
+
+	if writeErr != nil {
+		return writeErr
+	}
+
+	ok, err := fms.compareHash(file.GetFileMeta().GetName())
+	if !ok || err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (fms *FileManagerService) compareHash(filePath string) (bool, error) {
```

Comment:
What does it mean when `compareHash` returns `true`? I would consider renaming this function, as I don't think it's immediately obvious.

Reply:
Also, do we need to return the `bool` from `compareHash` at all, as we are always returning an error when the bool is `false`?

Could we just simply return an `error`?

Reply:
Changed it to return an error, it just checks what was written against the hash didn't really know what to name it, would `fileHashMatch` be better ?

Reply:
Yeah I like `fileHashMatch` more. Alternatively something like `checkFileHash` or `validateFileHash`, which tells the caller that we comparing the hash against a known valid hash.

Up to you though, just so long as it's less ambiguous than `compareHash` from the point of view of whoever is calling the function.